### ISSUE TITLE
Video FPS speed calculation not accurate

### ISSUE THREAD

**Issue by hargut** [alfredh/baresip#208]
With the latest improvements in the avcodec module for selecting h264 hardware en- & decoders I've discovered that baresips video engine within avformat has some limitations calculating the FPS.

The problematic code part is in the following area:
```
--- a/modules/avformat/avformat.c
+++ b/modules/avformat/avformat.c
@@ -180,7 +180,7 @@ static void handle_packet(struct vidsrc_st *st, AVPacket *pkt)
 #else
        dur = 1.0 / st->fps;
 #endif
-       sys_msleep(1000.0 * dur);
+//     sys_msleep(1000.0 * dur);
 
  out:
        if (frame) {
```
Below log uses 903ce9869c933e4f13995cd9f10c1498665cfd58 and the above patch in commit 2fb94e4d09f4fd36c4ad802fb6b908073dcede3e for comparison with a regular software h264 de-/encoder
```
hargut@hargut-laptop:~/bls-vte/baresip.hargut$ git checkout 903ce9869c933e4f13995cd9f10c1498665cfd58
Previous HEAD position was 2fb94e4... comment out speed decreasing line in avformat
HEAD is now at 903ce98... Elektm93 feat/configure snd file path (#207)
hargut@hargut-laptop:~/bls-vte/baresip.hargut$ make USE_X264="" EXTRA_MODULES="swscale"             
  CC [M]  build-x86_64/modules/avformat/avformat.o
  LD [M]  avformat.so
hargut@hargut-laptop:~/bls-vte/baresip.hargut$ ./baresip -v -e"/vidloop h264"
baresip v0.5.0 Copyright (C) 2010 - 2017 Alfred E. Heggestad et al.
Local network address:  IPv4=lxdbr0:10.10.10.29 
ui: stdio
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
aufilt: vumeter
ausrc: alsa
auplay: alsa
avcodec: using libavcodec H.264 encoder
vidcodec: H264
vidcodec: H263
vidcodec: MP4V-ES
vidfilt: swscale
vidsrc: v4l2
vidsrc: avformat
vidisp: sdl2
ausrc: rst
vidsrc: rst
medianat: stun
medianat: turn
medianat: ice
uuid: loaded UUID 5d42184c-34c8-11e8-4311-72f013d97d7f from file /home/hargut/.baresip/uuid
Populated 2 accounts
module: loading app auloop.so
module: loading app contact.so
Populated 3 contacts
module: loading app menu.so
module: loading app vidloop.so
Populated 2 audio codecs
Populated 1 audio filter
Populated 3 video codecs
Populated 1 video filter
baresip is ready.
/vidloop h264
Enable video-loop on avformat,http://10.10.10.130:3000/S19.2E-1-1019-10301.ts: 640 x 360
vidloop: added video-filter `swscale'
vidloop: avformat,http://10.10.10.130:3000/S19.2E-1-1019-10301.ts: open video source: 640 x 360 at 50 fps
[h264 @ 0x163e1a0] non-existing PPS 0 referenced
[h264 @ 0x163e1a0] non-existing PPS 0 referenced
[h264 @ 0x163e1a0] decode_slice_header error
[h264 @ 0x163e1a0] no frame!
[mpegts @ 0x1636200] Could not find codec parameters for stream 0 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options                                                                                                                                   
[mpegts @ 0x1636200] Could not find codec parameters for stream 1 (Unknown: none ([12][0][0][0] / 0x000C)): unknown codec                                                                                         
Consider increasing the value for the 'analyzeduration' and 'probesize' options                                                                                                                                   
[mpegts @ 0x1636200] Could not find codec parameters for stream 2 (Unknown: none ([11][0][0][0] / 0x000B)): unknown codec                                                                                         
Consider increasing the value for the 'analyzeduration' and 'probesize' options                                                                                                                                   
[mpegts @ 0x1636200] Could not find codec parameters for stream 10 (Unknown: none ([11][0][0][0] / 0x000B)): unknown codec                                                                                        
Consider increasing the value for the 'analyzeduration' and 'probesize' options                                                                                                                                   
avformat: stream 3:  1280 x 720   time_base=1/90000
vidloop: open video display (.)
vidloop: enabled encoder H264 (50 fps, 50000000 bit/s)
vidloop: enabled decoder H264
avcodec: video encoder H264: 50 fps, 50000000 bit/s, pktsize=1480
avcodec: video decoder H264 ()
Enabled codec: H264
swscale: created SwsContext: `yuv420p' 1280 x 720 --> `yuv420p' 640 x 360
[swscaler @ 0x7ff78c15bf80] Warning: data is not aligned! This can lead to a speedloss
[libx264 @ 0x7ff78c1ea400] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX
[libx264 @ 0x7ff78c1ea400] profile High, level 5.0
[h264 @ 0x1816260] cabac decode of qscale diff failed at 42 15    
[h264 @ 0x1816260] error while decoding MB 42 15, bytestream 24402
[h264 @ 0x1816260] concealing 2407 DC, 2407 AC, 2407 MV errors in B frame
[h264 @ 0x263b0a0] left block unavailable for requested intra mode at 0 12
[h264 @ 0x263b0a0] error while decoding MB 0 12, bytestream 139609
[h264 @ 0x263b0a0] concealing 2689 DC, 2689 AC, 2689 MV errors in I frame
Quitus: [H264] [H264]  intra=30  EFPS=16.4      6281 kbit/s       
ua: stop all (forced=0)
module: unloading app vidloop
[libx264 @ 0x7ff78c1ea400] frame I:32    Avg QP:10.05  size: 91774
[libx264 @ 0x7ff78c1ea400] frame P:281   Avg QP:10.05  size: 41244
[libx264 @ 0x7ff78c1ea400] mb I  I16..4:  4.2% 24.7% 71.1%
[libx264 @ 0x7ff78c1ea400] mb P  I16..4:  0.6%  1.9%  2.5%  P16..4: 40.9% 29.9% 23.6%  0.0%  0.0%    skip: 0.6%
[libx264 @ 0x7ff78c1ea400] final ratefactor: -8.32
[libx264 @ 0x7ff78c1ea400] 8x8 transform intra:28.6% inter:25.4%
[libx264 @ 0x7ff78c1ea400] coded y,uvDC,uvAC intra: 97.1% 98.1% 96.9% inter: 87.4% 73.7% 61.9%
[libx264 @ 0x7ff78c1ea400] i16 v,h,dc,p: 56%  4%  9% 31%
[libx264 @ 0x7ff78c1ea400] i8 v,h,dc,ddl,ddr,vr,hd,vl,hu: 30% 11% 14%  6%  6%  6%  7%  9% 10%
[libx264 @ 0x7ff78c1ea400] i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 27% 10%  7%  8%  9%  9%  9% 10% 10%
[libx264 @ 0x7ff78c1ea400] i8c dc,h,v,p: 52% 13% 29%  6%
[libx264 @ 0x7ff78c1ea400] Weighted P-Frames: Y:1.1% UV:1.1%
[libx264 @ 0x7ff78c1ea400] kb/s:18564.15
module: unloading app menu
menu: close (redial current_attempts=0)
module: unloading app contact
module: unloading app auloop
Stopping 2 useragents.. 
ua: sip-stack exit
ua exited -- stopping main runloop
main: unloading modules..
hargut@hargut-laptop:~/bls-vte/baresip.hargut$ git checkout 2fb94e4d09f4fd36c4ad802fb6b908073dcede3e
Previous HEAD position was 903ce98... Elektm93 feat/configure snd file path (#207)
HEAD is now at 2fb94e4... comment out speed decreasing line in avformat
hargut@hargut-laptop:~/bls-vte/baresip.hargut$ make USE_X264="" EXTRA_MODULES="swscale"                       
  CC [M]  build-x86_64/modules/avformat/avformat.o
modules/avformat/avformat.c: In function 'handle_packet':
modules/avformat/avformat.c:103:9: warning: variable 'dur' set but not used [-Wunused-but-set-variable]
  double dur;
         ^
  LD [M]  avformat.so
hargut@hargut-laptop:~/bls-vte/baresip.hargut$ ./baresip -v -e"/vidloop h264"
baresip v0.5.0 Copyright (C) 2010 - 2017 Alfred E. Heggestad et al.
Local network address:  IPv4=lxdbr0:10.10.10.29 
ui: stdio
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
aufilt: vumeter
ausrc: alsa
auplay: alsa
avcodec: using libavcodec H.264 encoder
vidcodec: H264
vidcodec: H263
vidcodec: MP4V-ES
vidfilt: swscale
vidsrc: v4l2
vidsrc: avformat
vidisp: sdl2
ausrc: rst
vidsrc: rst
medianat: stun
medianat: turn
medianat: ice
uuid: loaded UUID 5d42184c-34c8-11e8-4311-72f013d97d7f from file /home/hargut/.baresip/uuid
Populated 2 accounts
module: loading app auloop.so
module: loading app contact.so
Populated 3 contacts
module: loading app menu.so
module: loading app vidloop.so
Populated 2 audio codecs
Populated 1 audio filter
Populated 3 video codecs
Populated 1 video filter
baresip is ready.
/vidloop h264
Enable video-loop on avformat,http://10.10.10.130:3000/S19.2E-1-1019-10301.ts: 640 x 360
vidloop: added video-filter `swscale'
vidloop: avformat,http://10.10.10.130:3000/S19.2E-1-1019-10301.ts: open video source: 640 x 360 at 50 fps
[h264 @ 0x1960260] non-existing PPS 0 referenced
[h264 @ 0x1960260] non-existing PPS 0 referenced
[h264 @ 0x1960260] decode_slice_header error
[h264 @ 0x1960260] no frame!
[mpegts @ 0x19d4c60] Could not find codec parameters for stream 0 (Unknown: none ([5][0][0][0] / 0x0005)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[mpegts @ 0x19d4c60] Could not find codec parameters for stream 1 (Unknown: none ([12][0][0][0] / 0x000C)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[mpegts @ 0x19d4c60] Could not find codec parameters for stream 2 (Unknown: none ([11][0][0][0] / 0x000B)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
[mpegts @ 0x19d4c60] Could not find codec parameters for stream 10 (Unknown: none ([11][0][0][0] / 0x000B)): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
avformat: stream 3:  1280 x 720   time_base=1/90000
vidloop: open video display (.)
vidloop: enabled encoder H264 (50 fps, 50000000 bit/s)
vidloop: enabled decoder H264
avcodec: video encoder H264: 50 fps, 50000000 bit/s, pktsize=1480
avcodec: video decoder H264 ()
Enabled codec: H264
[h264 @ 0x20f8680] concealing 3012 DC, 3012 AC, 3012 MV errors in B frame
swscale: created SwsContext: `yuv420p' 1280 x 720 --> `yuv420p' 640 x 360
[swscaler @ 0x7f5afc15afe0] Warning: data is not aligned! This can lead to a speedloss
[libx264 @ 0x7f5afc1ea100] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX
[libx264 @ 0x7f5afc1ea100] profile High, level 5.0
[h264 @ 0x20d9640] concealing 2983 DC, 2983 AC, 2983 MV errors in P frame
[h264 @ 0x20d9640] left block unavailable for requested intra4x4 mode -1 at 0 7
[h264 @ 0x20d9640] error while decoding MB 0 7, bytestream 5425
[h264 @ 0x20d9640] concealing 3089 DC, 3089 AC, 3089 MV errors in I frame
[mpegts @ 0x19d4c60] PES packet size mismatch
[mpegts @ 0x19d4c60] PES packet size mismatch
[mpegts @ 0x19d4c60] PES packet size mismatch
Quitus: [H264] [H264]  intra=58  EFPS=35.4      7479 kbit/s       
ua: stop all (forced=0)
module: unloading app vidloop
[libx264 @ 0x7f5afc1ea100] frame I:61    Avg QP:10.02  size: 63336
[libx264 @ 0x7f5afc1ea100] frame P:541   Avg QP:10.02  size: 22796
[libx264 @ 0x7f5afc1ea100] mb I  I16..4:  6.0% 40.1% 53.9%
[libx264 @ 0x7f5afc1ea100] mb P  I16..4:  0.7%  2.3%  1.4%  P16..4: 52.1% 20.1% 16.3%  0.0%  0.0%    skip: 7.2%
[libx264 @ 0x7f5afc1ea100] final ratefactor: -13.03
[libx264 @ 0x7f5afc1ea100] 8x8 transform intra:43.4% inter:27.3%
[libx264 @ 0x7f5afc1ea100] coded y,uvDC,uvAC intra: 94.5% 96.6% 93.5% inter: 63.2% 54.8% 36.2%
[libx264 @ 0x7f5afc1ea100] i16 v,h,dc,p: 39% 13% 19% 28%
[libx264 @ 0x7f5afc1ea100] i8 v,h,dc,ddl,ddr,vr,hd,vl,hu: 23% 16% 17%  6%  7%  7%  8%  8%  9%
[libx264 @ 0x7f5afc1ea100] i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 18% 12%  8%  9% 12% 11% 10% 10%  9%
[libx264 @ 0x7f5afc1ea100] i8c dc,h,v,p: 55% 16% 24%  5%
[libx264 @ 0x7f5afc1ea100] Weighted P-Frames: Y:0.6% UV:0.6%
[libx264 @ 0x7f5afc1ea100] kb/s:10761.54
module: unloading app menu
menu: close (redial current_attempts=0)
module: unloading app contact
module: unloading app auloop
Stopping 2 useragents.. 
ua: sip-stack exit
ua exited -- stopping main runloop
main: unloading modules..
hargut@hargut-laptop:~/bls-vte/baresip.hargut$
```

Without the "sys_msleep(1000.0 * dur)" baresip is able to produce a stream with approx. 35 fps, with the line active and same source & configuration the stream will only have approx 17 fps. The input material is in this case a 50 fps 720p video stream.
```
//original source
Quitus: [H264] [H264]  intra=30  EFPS=16.4      6281 kbit/s
//above listed patch
Quitus: [H264] [H264]  intra=58  EFPS=35.4      7479 kbit/s
```
When commenting out that line & using hwenc/hwdec baresip can easily produce a 50 fps 720p stream live re-encoding. Activating that line will produce with hwenc/hwdec approx. 42 fps stream.
The calculation of the sleep time is not really accurate and approx. 15-25% off.

Reading files from local disk need some kind of speed-limitation, best would be to discover the native frame rate of the material (similar to ffmpeg's -re option). When handling streams the limitation will be the stream supply.

With the above patch, using hwenc / hwdec options reading 720p material from disk will produce a live re-encoded stream with approx. 80 fps.

Best regards,
Harald

**Comment by alfredh** [comment]
many thanks for the detailed report.

doing a blocking sleep is of course not a good solution.

what we should do instead:

1. read out the timestamp from `AVPacket.dts`
2. convert it to seconds using `av_q2d` and time_base
3. keep a local timer running using `tmr_jiffies`
4. compare timestamp from packet, if later than current time then decode and play packet
5. otherwise sleep for 4 milliseconds


an example of similar code which is working can be found
in the RST module:

https://github.com/alfredh/baresip/blob/master/modules/rst/audio.c#L58

**Comment by hargut** [comment]
Thank you for the suggestion. I've already had a look and implemented it, but the issue still needs some more tweaks & polish to be good.

I'm still not really sure how to handle the frame rate on the input side. Need to have a look at the details there. With the first try I ended up with 25 fps on 50 fps material, but with 25 fps it worked nicely, could be config related as well.
Generally I think we should at least try to read the fps from the file/stream instead reading some value from the config which might not even be close to the real fps rate. Probably we should use the fps value in the config as encoding fps & fallback value if we cannot determine it from the stream.

```st->dec_fps = av_q2d(strm->avg_frame_rate);```

In order to get that done I probably have to introduce avformat_probesize and avformat_analyzeduration to be able to handle network-streams with higher resolution in a proper way. Increasing these values might be necessary to have fps being detected.

**Comment by hargut** [comment]
Solved 029c4f60efbad9b1f0518ee2716195a86990ac9b
